### ISSUE TITLE
Implicit variables in groovysh

### ISSUE THREAD

**Issue by abhi18av** [mvmn/groovysh.rc#1]
Hi @mvmn  :)

You do seem to love groovy a lot, it's the only `groovysh.rc` that shows up on Google!

So, the thing is I'm trying to port a bunch of **perl one liners to groovy ** and I'm kinda stuck at a few points - especially with the implicit variables.

Like, there's 

* `line` and `count`  in `-n ` and `-p` 
* `split` in `-a`

I need to know whether there're implicit variables like 

- `word` for processing every single word per line or otherwise
- *and many more* (?)

Sadly, even the official doc. mentions only the `line` and `split` only and there are only a couple blog posts regarding this. I tried going through the source code but being a n00b, it's a bit overwhelming for me. I'd greatly appreciate your help 😄

**Comment by mvmn** [comment]
Hi

Groovy does not have implicit variables. Groovysh shell has some, but only for it's own internal workings.

If you want to do things like processing each word in a line you should do it in standard Java/Groovy way - split the line and apply foreach to each split.

Example:

```
line = "some text in a line"; // assign variable line to a string with one line of text
line.split("[ ]+").each({ println it; }); // split a line using a regex pattern - prouces an array of Strings,
// use Groovy .each feature to apply Groovy closure to each element of an array. 
// { println it; } is a closure with one parameter, 
// when not specified explicitly the parameter has name "it".
// println(stuff) in Groovy short for Java System.out.println(stuff), 
// parenthesis are optional in groovy, so "print stuff" works too.

```

**Comment by abhi18av** [comment]
@mvmn , is it possible to define some more constructs like `-n` and `-p` for the `groovysh`. How to do it i.e. where are they really defined in the source code? 

I went through the files but came across only a couple `bash` scripts. Could you point them out to me?

Would be awesome if there's a way to just define them using the `.groovyrc`

**Comment by mvmn** [comment]
I'm not sure what -n and -p is. It's something from perl?

**Comment by abhi18av** [comment]
Yes, [this blog](http://www.catonmat.net/blog/introduction-to-perl-one-liners/) might be a good introduction. The thing is even `python` doesn't have this and it's syntax quite unsuitable for the quick CLI programs. 

`Groovy` to my surprise already has most perl like switches in place!  And with `groovyserver/client` system - it's well suited for this work as well.
